clippy_lints/src/avoidable_slice_indexing.rs

```diff
@@ -0,0 +1,291 @@
+                let needs_ref = !bound_ty.is_ref()
+                    || !cx
+                        .tcx
+                        .lang_items()
+                        .copy_trait()
+                        .map_or(false, |trait_id| implements_trait(cx, inner_ty, trait_id, &[]));
+
+                let key = *ident_to_id.entry(ident.name).or_insert(pat.hir_id);
+                let slice_info = slices
+                    .entry(key)
+                    .or_insert_with(|| SliceLintInformation::new(ident, needs_ref));
+                slice_info.pattern_spans.push(pat.span);
+            }
+        }
+    });
+
+    (!slices.is_empty()).then(|| slices)
+}
+
+fn lint_slice(cx: &LateContext<'_>, slice: &SliceLintInformation) {
+    let used_indices = slice
+        .index_use
+        .iter()
+        .map(|(index, _)| *index)
+        .collect::<FxHashSet<_>>();
+
+    let value_name = |index| format!("{}_{}", slice.ident.name, index);
+
+    if let Some(max_index) = used_indices.iter().max() {
+        let opt_ref = if slice.needs_ref { "ref " } else { "" };
+        let pat_sugg_idents = (0..=*max_index)
+            .map(|index| {
+                if used_indices.contains(&index) {
+                    format!("{}{}", opt_ref, value_name(index))
+                } else {
+                    "_".to_string()
+                }
+            })
+            .collect::<Vec<_>>();
+        let pat_sugg = format!("[{}, ..]", pat_sugg_idents.join(", "));
+
+        span_lint_and_then(
+            cx,
+            AVOIDABLE_SLICE_INDEXING,
+            slice.ident.span,
+            "this slice can be deconstructured to avoid indexing",
+            |diag| {
+                diag.multipart_suggestion(
+                    "try destructing the slice with a pattern here",
+                    slice
+                        .pattern_spans
+                        .iter()
+                        .map(|span| (*span, pat_sugg.clone()))
+                        .collect(),
+                    Applicability::MaybeIncorrect,
+                );
+
+                diag.multipart_suggestion(
+                    "and replacing the index expressions here",
+                    slice
+                        .index_use
+                        .iter()
+                        .map(|(index, span)| (*span, value_name(*index)))
+                        .collect(),
+                    Applicability::MaybeIncorrect,
+                );
+
+                // I thought about adding a note to the lint message to inform the user that these
+                // refactorings will remove the expressions that determine which access should be
+                // used. However, I decided against this, as `filter_lintable_slices` will only
+                // return slices where all access indices are known at compile time. They will
+                // therefore not have any side effects when they are removed.
```

Comment:
Hmm what do you mean be "determine which access should be used"?

Reply:
This refers to the index expression. The new suggestion removes them completely (`slice[CONST]` → `slice_x`). However, this shouldn't have any side effects as the value is already constant.

I've changed the comment a bit to clearly state this :upside_down_face: